devito/types/dense.py

```diff
@@ -1360,6 +1361,31 @@ def __shape_setup__(cls, **kwargs):
                 shape.insert(cls._time_position, save)
             else:
                 raise TypeError("`save` can be None, int or Buffer, not %s" % type(save))
+        elif dimensions is None:
+            raise TypeError("`dimensions` required if both `grid` and "
+                            "`shape` are provided")
+        else:
+            # Got `grid`, `dimensions`, and `shape`. We sanity-check that the
```

Comment:
does it fail in some ways that I'm currently missing if we rather just do

```
else:
    return super().__shape_setup__(grid=grid, shape=shape, dimensions=dimensions)
```

Reply:
I don't think it fails without this PR, no. But, as things stand before this change:

*  `TimeFunction` is inconsistent with `Function`. The shape that must currently be provided for grid-owned `Dimension`s is the local shape not the global shape, whereas for `Function` it's the global shape (or None).
* it doesn't verify early that the provided shape is consistent with the `Grid` - you only get that error later, at `apply`/`arguments` time

Reply:
Oh, with my message above I really meant something else, sorry. I was rather proposing a potential way of avoiding a piece of duplicated code (at least, it seemed duplicated code at first glance).

Reply:
Ah, sorry, I get you now. Yes, good plan. Change pushed.